hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/type/ToRowJsonFunctionTest.java

```diff
@@ -0,0 +1,92 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.sql.impl.type;
+
+import com.hazelcast.config.Config;
+import com.hazelcast.core.HazelcastJsonValue;
+import com.hazelcast.jet.sql.SqlTestSupport;
+import com.hazelcast.map.IMap;
+import com.hazelcast.sql.HazelcastSqlException;
+import com.hazelcast.sql.impl.SqlErrorCode;
+import com.hazelcast.test.HazelcastSerialClassRunner;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import static com.hazelcast.jet.sql.impl.type.BasicNestedFieldsTest.createJavaMapping;
+import static com.hazelcast.jet.sql.impl.type.BasicNestedFieldsTest.createJavaType;
+import static com.hazelcast.spi.properties.ClusterProperty.SQL_CUSTOM_TYPES_ENABLED;
+import static java.lang.String.format;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+@RunWith(HazelcastSerialClassRunner.class)
+public class ToRowJsonFunctionTest extends SqlTestSupport {
+    @BeforeClass
+    public static void beforeClass() {
+        Config config = smallInstanceConfig()
+                .setProperty(SQL_CUSTOM_TYPES_ENABLED.getName(), "true");
+        initializeWithClient(3, config, null);
+    }
+
+    @Test
+    public void test_nonCyclic() {
+        client().getSql().execute(format("CREATE TYPE UserType (id BIGINT, name VARCHAR, organization OrganizationType) "
+                + "OPTIONS ('format'='java', 'javaClass'='%s')", BasicNestedFieldsTest.User.class.getName()));
+        client().getSql().execute(format("CREATE TYPE OrganizationType (id BIGINT, name VARCHAR, office OfficeType) "
+                + "OPTIONS ('format'='java', 'javaClass'='%s')", BasicNestedFieldsTest.Organization.class.getName()));
+        client().getSql().execute(format("CREATE TYPE OfficeType (id BIGINT, name VARCHAR) "
+                + "OPTIONS ('format'='java', 'javaClass'='%s')", BasicNestedFieldsTest.Office.class.getName()));
+        final IMap<Long, BasicNestedFieldsTest.User> testMap = client().getMap("test");
+        client().getSql().execute("CREATE MAPPING test (__key BIGINT, this UserType) "
+                + "TYPE IMap OPTIONS ("
+                + "'keyFormat'='bigint', "
+                + "'valueFormat'='java', "
+                + "'valueJavaClass'='" + BasicNestedFieldsTest.User.class.getName() + "')");
+
+        final BasicNestedFieldsTest.Office office = new BasicNestedFieldsTest.Office(3L, "office1");
+        final BasicNestedFieldsTest.Organization organization = new BasicNestedFieldsTest.Organization(2L, "organization1", office);
+        final BasicNestedFieldsTest.User user = new BasicNestedFieldsTest.User(1L, "user1", organization);
+        testMap.put(1L, user);
+
+        assertRowsAnyOrder("SELECT CAST(this AS JSON) FROM test", rows(1,
+                new HazelcastJsonValue("[1,\"user1\",[2,\"organization1\",[3,\"office1\"]]]")));
```

Comment:
Since we're casting only custom types, I expected the result to be a JSON object, not array. Custom types always have field names, unlike ROW values, which can't be converted to JSON. JSON object is more useful I think.